tests/privacy_simple_network.py

```diff
@@ -160,33 +163,120 @@ def verifyNonParticipants(transId):
+        verifyParticipantsTransactionFinalized(publishTransId)
+        verifyNonParticipants(publishTransId)
+
+    def moveToParticipants():
+        movedNode = nonParticipants[0]
+        participants.append(movedNode)
+        del nonParticipants[0]
+        return movedNode
+
+    def moveToNonParticipants():
+        movedNode = participants[-1]
+        # popping off back of participants and need to push on the front of nonParticipants
+        nonParticipants.insert(0, movedNode)
+        del participants[-1]
+        return movedNode
+
+    def addToSg():
+        node = moveToParticipants()
+        Utils.Print("Take a non-participant and make a participant. Now there are {} participants and {} non-participants".format(len(participants), len(nonParticipants)))
+        toAddNum = cluster.getParticipantNum(node)
+        return security_group([toAddNum])
+
+    def remFromSg():
+        node = moveToNonParticipants()
+        Utils.Print("Take a participant and make a non-participant. Now there are {} participants and {} non-participants".format(len(participants), len(nonParticipants)))
+        toRemoveNum = cluster.getParticipantNum(node)
+        return security_group(removeNodeNums=[toRemoveNum])
+
     Utils.Print("Add all producers to security group")
     publishTrans = security_group([x for x in range(pnodes)])
-    publishTransId = Node.getTransId(publishTrans[1])
-    verifyParticipantsTransactionFinalized(publishTransId)
-    verifyNonParticipants(publishTransId)
+    verifySecurityGroup(publishTrans)
+
+    cluster.reportInfo()
+
 
+    # one by one add each nonParticipant to the security group
     while len(nonParticipants) > 0:
-        toAdd = nonParticipants[0]
-        participants.append(toAdd)
-        del nonParticipants[0]
-        Utils.Print("Take a non-participant and make a participant. Now there are {} participants and {} non-participants".format(len(participants), len(nonParticipants)))
+        publishTrans = addToSg()
+        verifySecurityGroup(publishTrans)
+        cluster.reportInfo()
 
-        toAddNum = None
-        num = 0
-        for node in cluster.getNodes():
-            if node == toAdd:
-                toAddNum = num
-                break
-            num += 1
-        if toAddNum is None:
-            assert toAdd == cluster.biosNode
-            toAddNum = totalNodes
-        publishTrans = security_group([toAddNum])
-        publishTransId = Node.getTransId(publishTrans[1])
-        verifyParticipantsTransactionFinalized(publishTransId)
-        verifyNonParticipants(publishTransId)
 
+    # one by one remove each (original) nonParticipant from the security group
+    while len(participants) > pnodes:
+        publishTrans = remFromSg()
+        verifySecurityGroup(publishTrans)
+        cluster.reportInfo()
+
+
+    # if we have more than 1 api node, we will add and remove all those nodes in bulk, if not it is just a repeat of the above test
+    if totalNodes > pnodes + 1:
```

Comment:
you have `apiNodes` and `nonParticipants` list with all API nodes. I think it will be more clear to use here `len(apiNodes)` or `len(nonParticipants)`. btw, do you need two arrays for API nodes?